### ISSUE TITLE
Very Slow access to ALL_OBJECTS table in 12.2.0.1 container

### ISSUE THREAD

**Issue by davidritter** [oracle/docker-images#959]
Hello!

We are also seeing this issue using the Oracle Database 12c container. It seems like it is slow specifically when making queries against system tables, i.e.

`SELECT COUNT(*) FROM ALL_OBJECTS;`

Using that call against the container hosted database takes minutes to execute where against our non containerized Oracle server it returns practically instantly.

I think this issue should be re-opened. Please let me know what information I can provide to help diagnose it. To start with here is our docker and host OS information:

```
[ritter@larch] > sudo docker version
Client:
 Version:         1.13.1
 API version:     1.26
 Package version: <unknown>
 Go version:      go1.8.3
 Git commit:      774336d/1.13.1
 Built:           Wed Mar  7 17:06:16 2018
 OS/Arch:         linux/amd64

Server:
 Version:         1.13.1
 API version:     1.26 (minimum version 1.12)
 Package version: <unknown>
 Go version:      go1.8.3
 Git commit:      774336d/1.13.1
 Built:           Wed Mar  7 17:06:16 2018
 OS/Arch:         linux/amd64
 Experimental:    false
```

```
[ritter@larch] > sudo docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 31
Server Version: 1.13.1
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: journald
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: docker-runc runc
Default Runtime: docker-runc
Init Binary: docker-init
containerd version:  (expected: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1)
runc version: N/A (expected: 9df8b306d01f59d3a8029be411de015b7304dd8f)
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 seccomp
  WARNING: You're not using the default seccomp profile
  Profile: /etc/docker/seccomp.json
Kernel Version: 3.10.0-693.21.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 3
CPUs: 2
Total Memory: 3.702 GiB
Name: larch.cvo.roguewave.com
ID: UVTK:F7KU:JSRT:U42R:MPEI:MDPX:435N:L7DX:W3QO:AHIA:UWRI:7BZN
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Registries: docker.io (secure)
```

```
[ritter@larch] > cat /etc/*release*
CentOS Linux release 7.4.1708 (Core)
Derived from Red Hat Enterprise Linux 7.4 (Source)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.4.1708 (Core)
CentOS Linux release 7.4.1708 (Core)
cpe:/o:centos:centos:7
```

Our container is being run with this command:

`sudo docker run -d --restart always -e TZ=America/Los_Angeles -e ORACLE_PWD=<value> -p 1521:1521 -p 5500:5500 --name oracle-12.2.0.1  rw-oracle-12.2.0.1`

Where the image is just adding some databases and users using the setup directory and extending the image oracle/database:12.2.0.1-se2

I should also note I tried the
`ALTER SYSTEM SET disk_asynch_io = FALSE SCOPE = SPFILE;`
setting and did not observer that making any difference in execution time of the query.

One other note, I see that this is **ONLY** happening with the Oracle 12.2 container, the 12.1 container has normal speedy access to the ALL_OBJECTS table.

**Comment by gvenzl** [comment]
Unfortunately we cannot help much with why Docker appears to be slow on one machine while it runs fast on another. However, hopefully the community is of help here.

The one thing that I do see is that the Docker version is years old. I recommend upgrading to 18.09 and try again. Perhaps other people in the community have other tips or observations as well.

**Comment by Djelibeybi** [comment]
I would be interested to see you reproduce this issue on Oracle Linux 7 with the UEK4 and Oracle Container Runtime for Docker. All of those are far more recent than CentOS and are also free to download, distribute and use.

**Comment by davidritter** [comment]
I'm observing the same issue with the Oracle 18.3 container. This is my Dockerfile:

```
FROM oracle/database:18.3.0-se2

COPY initdb.d/* /docker-entrypoint-initdb.d/setup/
```

As you can see I am using the stock Oracle database container, which appears to be defined to use 
```
# Pull base image
# ---------------
FROM oraclelinux:7-slim as base
```
as it's OS.

Additionally I am now using Docker version 18.09
```
docker --version
Docker version 18.09.3, build 774a1f4
```

**Comment by davidritter** [comment]
The issue appears to continue with Oracle 19c

**Comment by abcoskar** [comment]
Any solution?

**Comment by Djelibeybi** [comment]
@gvenzl can you assign this to the DB folks who can look into this?

**Comment by gvenzl** [comment]
This has nothing to do with the Dockerfiles provided in this repository. Please file a ticket with Oracle Support instead.

**Comment by davidritter** [comment]
I just wanted to update this thread for anyone else that comes across it. I _think_ I have an understanding of what is going on. 

**TL;DR**
Add the SQL statement:
`ALTER SESSION SET CONTAINER = ORCLPDB1;`
before calls to create tablespaces and users so they will be created in the Pluggable Database container rather than the ROOT container

With the release of Oracle 12.2 Oracle [deprecated the non-CDB architecture](https://docs.oracle.com/en/database/oracle/oracle-database/12.2/upgrd/changes-in-12c-release-2-upgrade.html#GUID-85BD26D1-BE1B-4C72-A8BA-05D8797360D6). 

> * Deprecation of Non-CDB Architecture
>    The non-CDB architecture is deprecated in Oracle Database 12c release 1 (12.1), and may be desupported and unavailable in a release after Oracle Database 12c Release 2 (12.2).
>    Oracle recommends use of the CDB architecture.

With that release the `Dockerfile` and related code that creates the database in the image were updated to create databases as CDB databases rather than the deprecated non-CDB architecture. This makes sense, why keep using a deprecated feature.

But the side effect of this is if you don't know that the 12.2 release triggered a change in the underlying database architecture being used you won't realize that while the setup SQL files to create a new tablespace will continue to work without error (meaning using SQL scripts used with a previous version of Oracle) they will not being doing exactly what you want/need.

They will execute correctly, however the tablespace will be created in the `CDB$ROOT` container, rather than the `ORCLPDB1` container that is now available in the docker image for holding user data.

The practical upshot of all this is it appears that querying the ALL_OBJECTS (and possibly other views) from the ROOT container is very slow now. [This has been observed by other Oracle users in a non-Docker environment](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiIqMqK94zuAhWDrFkKHdJtB6gQFjAAegQIBhAC&url=https%3A%2F%2Fcommunity.oracle.com%2Fthread%2F4306711%3Fstart%3D0%26tstart%3D0&usg=AOvVaw2RZlMJEiioLB-mId5pEO3T). I imagine this may have to do with the ROOT container now needing to lock and query a lot more data to aggregate everything that is going on, but I'm by no means a CDB expert.

If however you use the `ORCLPDB1` container to host your tablespace and execute your queries logged in using that container they run with similar performance as previous versions (this also assumes you created local users in your PDB approrpriately now too)
